modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/fixed/CustomStaticTypeDeclarationEnricher.java

```diff
@@ -0,0 +1,163 @@
+import java.util.function.Function;
+
+/**
+ * Overrides the default output and input loaded types for the ones specified by the user using the custom static types features,
+ * such as {@link InputXmlType}, {@link OutputJsonType} and {@link InputStaticTypeResolver} for custom build types between others.
+ *
+ * @since 4.1
+ */
+public final class CustomStaticTypeDeclarationEnricher implements DeclarationEnricher {
+
+  @Override
+  public DeclarationEnricherPhase getExecutionPhase() {
+    return INITIALIZE;
+  }
+
+  @Override
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    new IdempotentDeclarationWalker() {
+
+      @Override
+      protected void onOperation(WithOperationsDeclaration owner, OperationDeclaration operation) {
+        operation.getModelProperty(ImplementingMethodModelProperty.class)
+            .map(ImplementingMethodModelProperty::getMethod)
+            .ifPresent(method -> getOutputType(method).ifPresent(type -> declareCustomType(operation.getOutput(), type)));
+
+        for (ParameterDeclaration param : operation.getAllParameters()) {
+          param.getModelProperty(ImplementingParameterModelProperty.class)
+              .map(ImplementingParameterModelProperty::getParameter)
+              .ifPresent(annotated -> getInputType(annotated).ifPresent(type -> declareCustomType(param, type)));
+        }
+      }
+
+      @Override
+      protected void onSource(WithSourcesDeclaration owner, SourceDeclaration source) {
+        source.getModelProperty(ImplementingTypeModelProperty.class)
+            .map(ImplementingTypeModelProperty::getType)
+            .ifPresent(clazz -> getOutputType(clazz).ifPresent(type -> declareCustomType(source.getOutput(), type)));
+      }
+    }.walk(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+  }
+
+  private <T extends BaseDeclaration & TypedDeclaration> void declareCustomType(T declaration, MetadataType overrideType) {
+    MetadataType type = declaration.getType();
+    Class<?> clazz = getType(type).orElseThrow(() -> new IllegalStateException("Could not find class in type [" + type + "]"));
+    Set<TypeAnnotation> annotations =
+        new HashSet<>(asList(new ClassInformationAnnotation(clazz), new CustomDefinedStaticTypeAnnotation()));
+    declaration.setType(new MetadataTypeEnricher().enrich(overrideType, annotations), false);
+  }
+
+  private Optional<MetadataType> getOutputType(AnnotatedElement element) {
+    OutputXmlType xml = element.getAnnotation(OutputXmlType.class);
+    if (xml != null) {
+      return of(parseXmlType(xml.schema(), xml.qName()));
+    }
+    OutputJsonType json = element.getAnnotation(OutputJsonType.class);
+    if (json != null) {
+      return of(parseJsonType(json.schema()));
+    }
+    OutputResolver resolver = element.getAnnotation(OutputResolver.class);
+    if (resolver != null && isStaticResolver(resolver.output())) {
+      return ofNullable(parseCustomStaticType(resolver.output()));
```

Comment:
`null` is not a valid return for the `StaticResolver`, I'd rather fail if we get a null there than ignoring it and leaving dead code there.

Reply:
I'm returning Optional to know if I should override the type or not. Changing ofNullable() -> of()